File: 2022/Sources/advent2022/Day8.swift
struct Day8 : Day {
    let sampleInput = [
        "30373",
        "25512",
        "65332",
        "33549",
        "35390"
    ]

    func parseLine (_ line :String) -> [Int] { line.map({ Int($0.asciiValue!) - Int(Character("0").asciiValue!) }) }
    func parseGrid (_ input :[String]) -> [[Int]] { input.map(parseLine) }
    func mapGrid<A> (_ grid :[[Int]], _ f :([[Int]], Int) -> A) -> [A] {
        func rays (_ row :Int, _ col :Int) -> [[Int]] {
            func slice (_ dr :Int, _ dc :Int) -> [Int] {
                var trees = [Int]()
                var r = row+dr ; var c = col+dc
                while (r >= 0 && c >= 0 && r < grid.count && c < grid[0].count) {
                    trees.append(grid[r][c])
                    r += dr
                    c += dc
                }
                return trees
            }
            return [slice(-1, 0), slice(1, 0), slice(0, -1), slice(0, 1)]
        }
        return grid.indices.flatMap { row in grid[row].indices.map { col in f(rays(row, col), grid[row][col]) }}
    }

    func part1 () throws -> String { String(mapGrid(parseGrid(try readInput(8)), {
        (rays, h) in rays.contains { $0.allSatisfy { $0 < h }}
    }).filter({ $0 }).count) }

    func part2 () throws -> String { String(mapGrid(parseGrid(try readInput(8)), {
        (rays, h) in rays.map({ 1 + ($0.firstIndex(where: { $0 >= h }) ?? $0.count-1) }).reduce(1,*)
    }).max() ?? 0) }
}

File: 2022/scala/src/main/scala/Day8.scala
object Day8 extends Day(8):

  def parseGrid (input :Seq[String]) = input.map(_.map(_ - '0'))

  def mapGrid[A] (grid :Seq[Seq[Int]])(f :(Array[Array[Int]], Int) => A) =
    def rays (row :Int, col :Int) =
      def slice (dr :Int, dc :Int) =
        val trees = collection.mutable.ArrayBuffer[Int]()
        var r = row+dr ; var c = col+dc
        while (r >= 0 && c >= 0 && r < grid.size && c < grid(0).size)
          trees.append(grid(r)(c))
          r += dr ; c += dc
        trees.toArray
      Array(slice(-1, 0), slice(1, 0), slice(0, -1), slice(0, 1))
    for (rr <- grid.indices ; cc <- grid(rr).indices) yield f(rays(rr, cc), grid(rr)(cc))

  override def answer1 (input :Seq[String]) =
    def isViz (rays :Array[Array[Int]], h :Int) = rays.exists(_.forall(_ < h))
    mapGrid(parseGrid(input))(isViz).filter(viz => viz).size

  override def answer2 (input :Seq[String]) =
    def viewDist (ray :Array[Int], h :Int) = ray.indexWhere(_ >= h) match
      case -1 => ray.size
      case  d => d+1
    mapGrid(parseGrid(input)) { (rays, h) => rays.map(r => viewDist(r, h)).product }.max
